### ISSUE TITLE
Update HuggingFace LoRA models with fixed 14-category versions

### ISSUE THREAD

**Issue by yossiovadia** [vllm-project/semantic-router#591]
## Summary

After PR #590 is merged (fixing the LoRA training regression), we need to train and upload new 14-category models to replace the broken 3-category models on HuggingFace.

## Background

Currently, the HuggingFace models under `LLM-Semantic-Router` organization only have 3 categories (business, law, psychology) instead of the required 14 MMLU-Pro categories. This affects all users who download pre-trained models via `make download-models`.

## Models That Need Updating

All three LoRA intent classifier models need to be retrained and uploaded:

1. `LLM-Semantic-Router/lora_intent_classifier_bert-base-uncased_model`
2. `LLM-Semantic-Router/lora_intent_classifier_roberta-base_model`
3. `LLM-Semantic-Router/lora_intent_classifier_modernbert-base_model`

## Training Parameters (Proven to Work)

Based on successful testing in PR #590, use these parameters:

```bash
python ft_linear_lora.py \
  --mode train \
  --model <base-model> \
  --max-samples 20000 \
  --epochs 12 \
  --batch-size 16 \
  --learning-rate 2e-5 \
  --lora-rank 16 \
  --lora-alpha 32
```

**Expected Results:**
- BERT: ~80% accuracy (validated)
- RoBERTa: ~78-82% accuracy (typically better than BERT)
- ModernBERT: ~75-80% accuracy

## Verification Checklist

Before uploading, verify each model has:
- ✅ 14 categories in `label_mapping.json` (all MMLU-Pro categories)
- ✅ Validation accuracy ≥75%
- ✅ Proper merged model in `*_rust` directory
- ✅ All required files: `model.safetensors`, `config.json`, `tokenizer.json`, etc.

## Impact

Until these models are updated:
- ⚠️ Users running `make download-models` get broken 3-category models
- ⚠️ Only users training their own models benefit from PR #590
- ⚠️ E2E tests may fail for fresh installations

## Categories Required (MMLU-Pro)

All models must support these 14 categories:
1. biology
2. business
3. chemistry
4. computer science
5. economics
6. engineering
7. health
8. history
9. law
10. math
11. other
12. philosophy
13. physics
14. psychology

## Prerequisites

- [ ] PR #590 merged
- [ ] Access to `LLM-Semantic-Router` HuggingFace organization
- [ ] GPU environment for training (takes ~30-40 min per model)

## Related

- Fixes remaining issues from #584
- Depends on #590

**Comment by Xunzhuo** [comment]
To run LoRA classification in VSR, I think maybe we need to do changes to VSR like API changes to support this instead of full params classification?

**Comment by yossiovadia** [comment]
Hi @Xunzhuo , the reason for this issue(#591) is that current exiting HuggingFace models are broken: All 3 models (BERT, RoBERTa, ModernBERT) only have *3 categories*

It affects all new users: Anyone running make download-models gets broken 3-category models
It also breaks E2E tests: Fresh installations will fail classification tests (30% accuracy instead of 75%+ as it was before) it seems someone / something re-uploaded the broken models and that caused the regression. 
Running the ft_linear_lora.py show a very bad accuracy ( 3% ) the FP32 fix mentioned below fixed it to be 75%.


What we need is - 
1. merge the FP32 PR (fixes training) https://github.com/vllm-project/semantic-router/pull/590 
2. Train new models with FP32 fix (generates correct 14-category models) , this is this issue we on now, #591
3. Upload to HuggingFace (fixes downloads for all users), ps I don't know who can/should do that. 

TBH - initially, i thought the regression occurred due to the LoRA refactor, but it's does not seems to be related. 

To your question, I would suggest to perhaps create a separate feature request ? ( I can do that if we agree)

**Comment by yossiovadia** [comment]
Current plan, also for future reference: 

Train command: 
---
1. BERT 
```python ft_linear_lora.py --mode train --model bert-base-uncased --max-samples 20000 --epochs 12 --batch-size 16 --learning-rate 2e-5 --lora-rank 16 --lora-alpha 32```

2. RoBERTa
```python ft_linear_lora.py --mode train --model roberta-base --max-samples 20000 --epochs 12 --batch-size 16 --learning-rate 2e-5 --lora-rank 16 --lora-alpha 32```

3. ModernBERT
```python ft_linear_lora.py --mode train --model modernbert-base --max-samples 20000 --epochs 12 --batch-size 16 --learning-rate 2e-5 --lora-rank 16 --lora-alpha 32```
---
Validation: 

```
# Check the _rust merged model was created
ls -la lora_intent_classifier_bert-base-uncased_r16_rust/
ls -la lora_intent_classifier_roberta-base_r16_rust/
ls -la lora_intent_classifier_modernbert-base_r16_rust/

# Verify 14 categories in config.json, e.g. 
cat lora_intent_classifier_bert-base-uncased_r16_rust/config.json | python3 -c "import sys, json; d=json.load(sys.stdin); print('Categories:', len(d.get('id2label', {})))"
```
Should output: Categories: 14


Then upload to HF.

**Comment by yossiovadia** [comment]
Results of each training: 
bert:
```
(.venv) ❖ WINDOWS yovadia@yos ~/semantic_router/src/training/training_lora/classifier_model_fine_tuning_lora # python ft_linear_lora.py --mode train --model bert-base-uncased --max-samples 20000 --epochs 12 --batch-size 16 --learning-rate 2e-5 --lora-rank 16 --lora-alpha 32
...
2025-11-06 08:23:42 - common_lora_utils - INFO - Validation Results:
2025-11-06 08:23:42 - common_lora_utils - INFO -   Accuracy: 0.7782
2025-11-06 08:23:42 - common_lora_utils - INFO -   F1: 0.7747
(.venv) ❖ WINDOWS yovadia@yos ~/semantic_router/src/training/training_lora/classifier_model_fine_tuning_lora # cat lora_intent_classifier_bert-base-uncased_r16_rust/config.json | python3 -c "import sys, json; d=json.load(sys.stdin); print('Categories:', len(d.get('id2label', {})))"

Categories: 14
```

roberta:
```
(.venv) ❖ WINDOWS yovadia@yos ~/semantic_router/src/training/training_lora/classifier_model_fine_tuning_lora # python ft_linear_lora.py --mode train --model roberta-base --max-samples 20000 --epochs 12 --batch-size 16 --learning-rate 2e-5 --lora-rank 16 --lora-alpha 32
2025-11-06 08:49:39 - common_lora_utils - INFO - Validation Results:
2025-11-06 08:49:39 - common_lora_utils - INFO -   Accuracy: 0.8119
2025-11-06 08:49:39 - common_lora_utils - INFO -   F1: 0.8112
(.venv) ❖ WINDOWS yovadia@yos ~/semantic_router/src/training/training_lora/classifier_model_fine_tuning_lora # cat lora_intent_classifier_roberta-base_r16_rust/config.json | python3 -c "import sys, json; d=json.load(sys.stdin); print('Categories:', len(d.get('id2label', {})))"
                                                                                                                                                                                                                      Categories: 14
```

For some reason, running the same training command for ModernBert, i get 
```
2025-11-06 08:58:56 - common_lora_utils - INFO - Model will be saved to: lora_intent_classifier_modernbert-base_r16
2025-11-06 08:58:56 - common_lora_utils - INFO - Starting training...
  0%|                                                                                                                                                                                        | 0/5784 [00:00<?, ?it/s]{'loss': 0.0, 'grad_norm': nan, 'learning_rate': 5.172413793103449e-07, 'epoch': 0.02}
{'loss': 0.0, 'grad_norm': nan, 'learning_rate': 1.0919540229885058e-06, 'epoch': 0.04}                                                                                                                               {'loss': 0.0, 'grad_norm': nan, 'learning_rate': 1.6666666666666667e-06, 'epoch': 0.06}
```

The NaN is what we saw with the FP16. Will dig

**Comment by yossiovadia** [comment]
I was unable to resolve the above. 
however, i uploaded to HF an existing modernbet model  from category_classifier_modernbert-base_model which had 14 categories.

test are not passing better ( 7/10 ) 
I'll adapt the 3 failed one on separate issue.
